### PR TITLE
Social Graph

### PR DESCRIPTION
**Pull request by arkin0x**
Converted social graph to typescript. Haven't tested this yet because I can't run the app :/

# TODO

- [ ] test in app
- [ ] database storage of social graph, database provider --- need to review how it works and implement
- [ ] write tests
- [ ] implement "web of distrust"

# Overview

Once the ArcadeSocial class is initialized it will start with the user's contact list and steadily branch out and build a social graph of connections until it has reached a depth of 3. It does not increase depth until all contacts on the current depth have been indexed (breadth first).

After depth 3 has been fully indexed, it will simply start over, iterating over every user in the graph and re-indexing users who are stale (1 week since last update). If a user is not stale, it waits 1 second before continuing its evaluation to not block any threads.

I don't think this will be too CPU intensive. It moves as fast as the relays return info. In testing (before porting to TS) it seemed to have a lot of downtime between relay responses.

### CODE REVIEW

File: src/social.ts [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+  }
+  /**
+   * Provide a public key to extend the social graph by one degree. Hey that rhymes!
+   * @param pubkey PublicKey - raw hex public key
+   * @param degree the current degree of separation from the user in this iteration
+   */
+  extendGraph(pubkey: PublicKey, degree = 1): void {
+    const kind3: EventTemplate[] = [];
+    const filter: Filter<number> = { kinds: [3], authors: [pubkey] };
+    this.pool.sub([filter], event => kind3.push(event), () => {
```
Comment (earonesty) [inline src/social.ts:62]: you're gonna need kind0 as well.   most clients stick the contacts in tags in kind0 instead.

**Comment by earonesty** [comment]
poing me when you added a test suite, its a lot easier to reason about this stuff when there's tests.

File: src/social.ts (hunk 1 above)
Comment (arkin0x) [reply src/social.ts:62]: I've been having trouble finding an example of this. Do they put the contacts in the tags the same way as kind 3, but just in a kind 0 event?

**Comment by arkin0x** [comment]
@earonesty I've just pushed some tests and they are passing. I'm not sure if they are as comprehensive as you would like them to be but I can add more.

File: src/social.ts [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+  //     .map(s => this.socialGraph[s])
+  //     .sort((a, b) => b.fwf - a.fwf)
+  //     .slice(0, 10)
+  //     .map(s => ({ pubkey: s.pubkey, fwf: s.fwf }))
+  //   console.log('top 10 fwf contacts', top10)
+
+  // }
+  // given a pubkey, return the pubkeys between it and the root or an empty array if none
+  // the pubkeys will be in the order of closest to your pubkey to closest to the target pubkey
+  traverse(pubkey: string) {
```
Comment (earonesty) [inline src/social.ts:231]:
```
// can you do these, and add to test?

distance(pubkey) {
   if (pubkey == root) return 0;
   if (pubkey not in this graph) return high_number;
   if (pubkey in this graph) traverse(pubkey).length) + 1
}

weight(pubkey) {
   if distance == 0l return high_numberl
   return 1 / distance * distance;
}
```

Comment (arkin0x) [reply src/social.ts:231]: I was unable to push to the social-graph branch anymore... permissions changed? Here is my new PR from a fork: https://github.com/ArcadeLabsInc/arclib/pull/36 @earonesty

**Comment by arkin0x** [comment]
See https://github.com/ArcadeLabsInc/arclib/pull/36

**Pull request closed without merging**